x86/src/ops/math.rs

```diff
@@ -4,49 +4,154 @@ use iced_x86::{Instruction, Register};
 use memory::Mem;
 use num_traits::ops::overflowing::OverflowingSub;
 
-/// This trait is implemented for u32/u16/u8 and lets us write operations generically
+/// This trait is implemented for u64/i64/u32/i32/u16/i16/u8/i8 and lets us write operations generically
 /// over all those bit sizes.
 ///
 /// Even when we need size-specific masks like "the high bit"
 /// (which is x.shr(I::bits() - 1))
 /// that math optimizes down to the appropriate constant.
 pub(crate) trait Int: num_traits::PrimInt {
-    fn as_usize(self) -> usize;
     fn bits() -> usize;
+
+    type Wide: num_traits::PrimInt;
+    fn widen(self) -> Self::Wide;
+    fn from_wide(wide: Self::Wide) -> Self;
 }
 impl Int for u64 {
-    fn as_usize(self) -> usize {
-        unimplemented!()
-    }
     fn bits() -> usize {
         64
     }
+    type Wide = u128;
+    fn widen(self) -> Self::Wide {
+        self as u128
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as u64
+    }
 }
 impl Int for u32 {
-    fn as_usize(self) -> usize {
-        self as usize
-    }
     fn bits() -> usize {
         32
     }
+    type Wide = u64;
+    fn widen(self) -> Self::Wide {
+        self as u64
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as u32
+    }
 }
 impl Int for u16 {
-    fn as_usize(self) -> usize {
-        self as usize
-    }
     fn bits() -> usize {
         16
     }
+    type Wide = u32;
+    fn widen(self) -> Self::Wide {
+        self as u32
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as u16
+    }
 }
 impl Int for u8 {
-    fn as_usize(self) -> usize {
-        self as usize
+    fn bits() -> usize {
+        8
     }
+    type Wide = u16;
+    fn widen(self) -> Self::Wide {
+        self as u16
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as u8
+    }
+}
+impl Int for i64 {
+    fn bits() -> usize {
+        64
+    }
+    type Wide = i128;
+    fn widen(self) -> Self::Wide {
+        self as i128
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as i64
+    }
+}
+impl Int for i32 {
+    fn bits() -> usize {
+        32
+    }
+    type Wide = i64;
+    fn widen(self) -> Self::Wide {
+        self as i64
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as i32
+    }
+}
+impl Int for i16 {
+    fn bits() -> usize {
+        16
+    }
+    type Wide = i32;
+    fn widen(self) -> Self::Wide {
+        self as i32
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as i16
+    }
+}
+impl Int for i8 {
     fn bits() -> usize {
         8
     }
+    type Wide = i16;
+    fn widen(self) -> Self::Wide {
+        self as i16
+    }
+    fn from_wide(wide: Self::Wide) -> Self {
+        wide as i8
+    }
+}
+
+pub(crate) trait UnsignedInt: Int {
+    fn as_usize(self) -> usize;
+}
+
+impl UnsignedInt for u64 {
+    fn as_usize(self) -> usize {
+        self as usize
```

Comment:
This is a truncate under wasm, because it goes from u64->u32. 
I think maybe the best solution is to just leave out the impl for u64, if that works?

Reply:
It looks like the only place .as_usize is used is in the impls of shifting, so maybe a better name for the method would be .as_shift() or something?  That would also make it clearer which numbers actually need it.

Reply:
Now that I'm using `overflowing_mul` I didn't need to touch these traits at all, so dropped this commit